core/src/main/scala/better/files/Implicits.scala

```diff
@@ -141,27 +142,51 @@ trait Implicits {
     /**
      * Provides and iterator that closes the underlying resource
      * when done.
-     * 
+     *
      * e.g.
      * <pre>
      * inputStream.autoClosedIterator(_.read())(_ != -1).map(_.toByte)
      * </pre>
      *
-     * @param next next element in this resource
-     * @param hasNext a function which tells if there is no more B left
+     * @param generator next element from this resource
+     * @param isValidElement a function which tells if there is no more B left e.g. certain iterators may return nulls
      * @tparam B
      * @return An iterator that closes itself when done
      */
-    def autoClosedIterator[B](next: A => B)(hasNext: B => Boolean): Iterator[B] = {
+    def autoClosedIterator[B](generator: A => B)(isValidElement: B => Boolean): Iterator[B] = {
       var isClosed = false
       def isOpen(item: B) = {
-        if(!isClosed && !hasNext(item)) {
-          resource.close()
-          isClosed = true
-        }
+        if(!isClosed && !isValidElement(item)) close()
         !isClosed
       }
-      Iterator.continually(next(resource)).takeWhile(isOpen)
+
+      def close() = {
+        resource.close()
+        isClosed = true
+      }
+
+      def next() = try {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Default case not specified.](https://www.codacy.com/app/pathikrit/better-files/file/1938297994/issues/source?bid=2860326&fileBranchId=3103914#l168)